### PR TITLE
Initial implementation of GPUDevice for WebGPU

### PR DESCRIPTION
**Pull request by imiklos**
Added the WebIDL bindigs for GPUDevice, GPUObjectDescriptorBase, GPUDeviceDescriptor, GPUObjectBase
Implemented the `requestDevice` function of `GPUAdapter`

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X]  These changes addresses a part of #24706

cc @jdm, @kvark, @zakorgy

### CODE REVIEW

File: components/webgpu/lib.rs [hunk 1]
```diff
@@ -77,6 +83,7 @@ struct WGPU {
     adapters: Vec<WebGPUAdapter>,
     // Track invalid adapters https://gpuweb.github.io/gpuweb/#invalid
     _invalid_adapters: Vec<WebGPUAdapter>,
+    device_to_adapter: HashMap<WebGPUDevice, WebGPUAdapter>,
```
Comment (kvark) [inline components/webgpu/lib.rs]: do we need this map?
it seems like we should just carry a JS reference to `WebGPUAdapter` from the device struct

**Review by kvark — CHANGES_REQUESTED** [review]

File: components/webgpu/lib.rs (hunk 1 above)
Comment (imiklos) [reply components/webgpu/lib.rs]: We don't use it right now. I will remove this map. The reference to the `Adapter` is already in place. I will update the pull request with this change. Thank you!

**Review by kvark — APPROVED** [review]

File: components/script/dom/gpudevice.rs [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    #[ignore_malloc_size_of = "mozjs"]
+    extensions: Heap<*mut JSObject>,
+    #[ignore_malloc_size_of = "mozjs"]
+    limits: Heap<*mut JSObject>,
+    label: DomRefCell<Option<DOMString>>,
+    device: WebGPUDevice,
+}
+
+impl GPUDevice {
+    pub fn new_inherited(
```
Comment (jdm) [inline components/script/dom/gpudevice.rs]: nit: no need for pub.

File: components/script/dom/gpuadapter.rs [hunk 3]
```diff
@@ -60,4 +71,47 @@ impl GPUAdapterMethods for GPUAdapter {
...
+
+impl AsyncWGPUListener for GPUAdapter {
+    fn handle_response(&self, response: WebGPUResponse, promise: &Rc<Promise>) {
+        match response {
+            WebGPUResponse::RequestDevice(device_id, _descriptor) => {
+                let device = GPUDevice::new(
+                    &self.global(),
+                    &self,
+                    Heap::default(),
+                    Heap::default(),
```
Comment (jdm) [inline components/script/dom/gpuadapter.rs]: Will we create non-default values based on descriptor in the future?

File: components/script/dom/webidls/GPUDevice.webidl [hunk 4]
```diff
@@ -0,0 +1,31 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+// https://gpuweb.github.io/gpuweb/#gpudevice
+[Exposed=(Window, DedicatedWorker)/*, Serializable */]
```
Comment (jdm) [inline components/script/dom/webidls/GPUDevice.webidl]: This should be `Pref="dom.webgpu.enabled"`.

**Review by jdm — CHANGES_REQUESTED** [review]

File: components/script/dom/gpuadapter.rs [hunk 5]
```diff
@@ -60,4 +71,47 @@ impl GPUAdapterMethods for GPUAdapter {
     fn Extensions(&self, _cx: SafeJSContext) -> NonNull<JSObject> {
         NonNull::new(self.extensions.get()).unwrap()
     }
+
+    /// https://gpuweb.github.io/gpuweb/#dom-gpuadapter-requestdevice
+    fn RequestDevice(&self, descriptor: &GPUDeviceDescriptor, comp: InCompartment) -> Rc<Promise> {
+        let promise = Promise::new_in_current_compartment(&self.global(), comp);
+        let sender = response_async(&promise, self);
+        let id = self.global().as_window().Navigator().create_device_id();
```
Comment (CYBAI) [inline components/script/dom/gpuadapter.rs]: I just noticed that `GPUAdapter` is possibly exposed to `DedicatedWorker`.

Maybe we should call these window methods inside a `if let` instead 👀?

Ex.

```rust
if let Some(window) = self.global().downcast::<Window>() {
    let id = window.Navigator().create_device_id();
    ...
}
```

File: components/script/dom/gpuadapter.rs (hunk 3 above)
Comment (imiklos) [reply components/script/dom/gpuadapter.rs]: Yes, we will deal with it in a follow up patch.

File: components/script/dom/gpuadapter.rs (hunk 5 above)
Comment (imiklos) [reply components/script/dom/gpuadapter.rs]: Thank you! I will updated the pull request with this change.

**Comment by imiklos** [comment]
@jdm @CYBAI Thank you for your review! I've addressed your comments and updated the PR.

**Comment by kvark** [comment]
@jdm error looks unrelated to the PR (android release font linking)

**Comment by jdm** [comment]
@bors-servo retry
* #24715

**Comment by jdm** [comment]
@bors-servo retry
* network issue

**Comment by jdm** [comment]
@bors-servo retry
* #24726 #21067

**Pull request merged**